core/embed/rust/src/ui/model_tt/component/keyboard/passphrase.rs

```diff
@@ -313,14 +350,140 @@ struct Input {
     area: Rect,
     textbox: TextBox,
     multi_tap: MultiTapKeyboard,
+    display_style: DisplayStyle,
+    last_char_timer: Timer,
 }
 
 impl Input {
+    const TWITCH: i16 = 4;
+    const X_STEP: i16 = 12;
+    const Y_STEP: i16 = 5;
+
     fn new() -> Self {
         Self {
             area: Rect::zero(),
             textbox: TextBox::empty(MAX_LENGTH),
             multi_tap: MultiTapKeyboard::new(),
+            display_style: DisplayStyle::LastChar,
+            last_char_timer: Timer::new(),
+        }
+    }
+
+    fn set_display_style(&mut self, display_style: DisplayStyle) {
+        self.display_style = display_style;
+    }
+
+    fn start_timer(&mut self, ctx: &mut EventCtx, timeout_ms: u32) {
+        self.last_char_timer
+            .start(ctx, Duration::from_secs(timeout_ms));
+    }
+
+    fn stop_timer(&mut self) {
+        self.last_char_timer.stop();
+    }
+
+    fn render_chars<'s>(&self, area: Rect, target: &mut impl Renderer<'s>) {
```

Comment:
If I look correctly, it seems that majority of the code in `render_chars` and `render_dots` is the same except the final rendering and the naming of variables (`chars` vs `all_chars`, `text_baseline` vs `cursor`). 

Do you think we can somehow simplify these two functions?

Reply:
https://github.com/trezor/trezor-firmware/pull/4385/commits/dc65c00243fd62224235819c6f5c05622411a7e7